jax/experimental/jax2tf/tests/tf_test_util.py

```diff
@@ -197,49 +125,100 @@ def ConvertAndCompare(self, func_jax: Callable, *args,
+        expect a failure.
+      custom_assert: a function that will be called `custom_assert(result_jax,
+        result_tf)` to assert equality of the results. The result_tf has been
+        converted to numpy. Use this function when
+        JAX and TF produce different results. This function is only used for
+        "eager" and "graph" modes by default, not for the "compiled" mode,
+        because in that case we expect the results to
         be equal (default: None).
       always_custom_assert: if True, custom_assert is also called in "compiled"
         mode. This is useful in cases where JAX and TF produce different but
         equally valid results (default: False).
-      enable_xla: if True, allows the use of XLA ops in jax2tf.convert
-        (default: True).
     """
-    # Run JAX
-    result_jax = func_jax(*args)
-    # Run TF in all execution modes
-    func_tf = jax2tf.convert(func_jax, enable_xla=enable_xla)
+    # Run JAX. Should not fail, we assume that the harness has been filtered
+    # already by JAX unimplemented primitives.
+    result_jax = func_jax(*args)  # JAX
+    result_tf = None
 
+    func_tf = jax2tf.convert(func_jax, enable_xla=enable_xla)
     tf_args = _make_tf_args(args)
 
-    for mode in ("eager", "graph", "compiled"):
-      result_tf = self.RunAndDetectLimitations(func_tf, *tf_args, mode=mode)
-      if result_tf is None:
-        continue
-
-      if custom_assert is not None and (mode in ("eager", "graph") or
-                                        always_custom_assert):
-        # If we have a custom assert, use it even if we expect incorrect results
-        custom_assert(result_jax, result_tf)
+    unexpected_successes = []
+    for mode in ("compiled", "eager", "graph"):
+      try:
+        result_tf = _run_tf_function(func_tf, *tf_args, mode=mode)
+        tf_exception = None
+      except Exception as e:
+        tf_exception = e
+
+      if limitations is not None:
+        dut = jtu.device_under_test()
+        lookup_mode = mode if dut != "tpu" else "compiled"
+        jax2tf_limits = limitations(dut, lookup_mode)
+      else:
+        jax2tf_limits = []
+
+      disable_comparison = any([l.disable_comparison for l in jax2tf_limits])
+      jax2tf_limits = [l for l in jax2tf_limits if not l.disable_comparison]
+      if tf_exception:
+        if jax2tf_limits:
+          logging.info(
+              f"[{self._testMethodName}] Found expected TF {mode} failure "
+              f" {tf_exception}; enabled limitations {[l.description for l in jax2tf_limits]}"
+          )
+          continue
+        else:
+          raise tf_exception
+      else:
+        if jax2tf_limits:
+          # It is more ergonomic to print all successful modes once
+          msg = (f"mode: {mode}; enabled limitations "
+                 f"{[l.description for l in jax2tf_limits]}")
+          unexpected_successes.append(msg)
+          logging.warning(f"Unexpected successful mode: {msg}")
```

Comment:
On the one hand, I would like to raise an exception here; on the other hand, it might be too burdensome as it will cause test failures for updates to TensorFlow. I assume this is why you chose not to do this.

Reply:
I decided that I do not want to burden people fixing TF/XLA. They would have to make the change in OSS at the same time as the change in google3.